### PR TITLE
Enhancement/logger factory

### PR DESCRIPTION
**Pull request by kezhenxu94**
## What is the purpose of the change

Polish/Enhance code to avoid many nested `try-catch` blocks

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by CrazyHZM** [comment]
Looks good to me.

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java [hunk 1]
```diff
@@ -34,40 +36,44 @@
...
-                        } catch (Throwable e4) {
-                            setLoggerAdapter(new JdkLoggerAdapter());
-                        }
+                        setLoggerAdapter(clazz.newInstance());
+                        break;
+                    } catch (Throwable ignored) {
                     }
                 }
-            }
+                break;
```
Comment (satansk) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: Seems no need to `break` in default ?

Comment (kezhenxu94) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: yes, just a personal habit, I've removed it

**Comment by kezhenxu94** [comment]
@satansk unnecessary `break` in `switch` block has been removed and the ci tests failed on Java11, which seems unrelated to this PR, mind having a look?

**Comment by satansk** [comment]
The CI don't work correctly sometimes, maybe we should provide some command like `rebuild` or `rerun` to deal with it, but for now, I think you can use an empty commit to trigger another CI :)

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java [hunk 2]
```diff
@@ -34,40 +36,43 @@
...
+                setLoggerAdapter(new JdkLoggerAdapter());
+                break;
+            case "log4j2":
+                setLoggerAdapter(new Log4j2LoggerAdapter());
+                break;
+            default:
+                List<Class<? extends LoggerAdapter>> candidates = Arrays.asList(
+                        Log4jLoggerAdapter.class,
+                        Slf4jLoggerAdapter.class,
+                        Log4j2LoggerAdapter.class,
```
Comment (cvictory) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: Is it need 3 other log class:  Log4j2LoggerAdapter, JclLoggerAdapter, JdkLoggerAdapter.

Comment (kezhenxu94) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: image
What do you mean, did I miss anything out?

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java [hunk 3]
```diff
@@ -103,12 +108,7 @@ public static void setLoggerAdapter(LoggerAdapter loggerAdapter) {
      * @return logger
      */
     public static Logger getLogger(Class<?> key) {
-        FailsafeLogger logger = LOGGERS.get(key.getName());
-        if (logger == null) {
-            LOGGERS.putIfAbsent(key.getName(), new FailsafeLogger(LOGGER_ADAPTER.getLogger(key)));
-            logger = LOGGERS.get(key.getName());
-        }
-        return logger;
+        return LOGGERS.computeIfAbsent(key.getName(), name -> new FailsafeLogger(LOGGER_ADAPTER.getLogger(key)));
```
Comment (ralf0131) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: I would suggest :

```java
return LOGGERS.computeIfAbsent(key.getName(), name -> new FailsafeLogger(LOGGER_ADAPTER.getLogger(name)));
```
for a better readability.
Note parameter for `getLogger` has been changed from `key` to `name`.

Comment (kezhenxu94) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: @ralf0131 If that's the case, why not just call the overloaded version `public static Logger getLogger(String key)`

Comment (ralf0131) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: When you are writing lambda x -> f(x) , `x` is just a placeholder, but in the function f(x) you treat  x as input parameters. 

Equivalently, in your case, it should be `name -> new FailsafeLogger(LOGGER_ADAPTER.getLogger(name))`, and `key` will be passed to the function at runtime.

You've implemented the same logic at Line 121.

When you are writing `name -> new FailsafeLogger(LOGGER_ADAPTER.getLogger(key))`, you are treating the function as zero parameter function, where `key` is a constant in this case.

Comment (kezhenxu94) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: I know what you meant, I'll update this patch to show what I meant

Comment (kezhenxu94) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: @ralf0131 Updated.

Comment (ralf0131) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: No, I didn't mean to call the overloaded method. Instead, I prefer your initial implementation. Call the overloaded method will introduce more overhead.

Your initial implementation is logically correct, what I mean is a minor change to improve the readability. You just need to replace `key` to `name`.

Comment (kezhenxu94) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: > You just need to replace key to name
That will also change the logic before this patch, it assumes that `org.apache.dubbo.common.logger.LoggerAdapter#getLogger(java.lang.Class<?>)` and `org.apache.dubbo.common.logger.LoggerAdapter#getLogger(java.lang.String)` return the same logger(which may be true in this case).

My thought was that `org.apache.dubbo.common.logger.LoggerFactory#getLogger(java.lang.Class<?>)` will get the logger via the key which is type of `Class` and, if I "replace key with name", the actual logger is got via name(which is type of `String`), anyway, it's the same in this situation, I'll update it soon

Comment (kezhenxu94) [reply dubbo-common/src/main/java/org/apache/dubbo/common/logger/LoggerFactory.java]: @ralf0131 updated

**Comment by ralf0131** [comment]
Hi, it looks good to me!

I see the unit test for `org.apache.dubbo.common.logger.LoggerFactory#getLogger(java.lang.String)` is missing, would you mind add it as well?

**Comment by kezhenxu94** [comment]
@ralf0131 added

**Pull request merged**